src/iop/diffuse.c

```diff
@@ -0,0 +1,1481 @@
+
+// The B spline best approximate a Gaussian of standard deviation :
+// see https://eng.aurelienpierre.com/2021/03/rotation-invariant-laplacian-for-2d-grids/
+#define B_SPLINE_SIGMA 1.0553651328015339f
+
+static inline float normalize_laplacian(const float sigma)
+{
+  // Normalize the wavelet scale to approximate a laplacian
+  // see https://eng.aurelienpierre.com/2021/03/rotation-invariant-laplacian-for-2d-grids/#Scaling-coefficient
+  return 2.f * M_PI_F / (sqrtf(M_PI_F) * sqf(sigma));
+}
+
+static inline float equivalent_sigma_at_step(const float sigma, const unsigned int s)
+{
+  // If we stack several gaussian blurs of standard deviation sigma on top of each other,
+  // this is the equivalent standard deviation we get at the end (after s steps)
+  // First step is s = 0
+  // see
+  // https://eng.aurelienpierre.com/2021/03/rotation-invariant-laplacian-for-2d-grids/#Multi-scale-iterative-scheme
+  if(s == 0)
+    return sigma;
+  else
+    return sqrtf(sqf(equivalent_sigma_at_step(sigma, s - 1)) + sqf(exp2f((float)s) * sigma));
+}
+
+static inline unsigned int num_steps_to_reach_equivalent_sigma(const float sigma_filter, const float sigma_final)
+{
+  // The inverse of the above : compute the number of scales needed to reach the desired equivalent sigma_final
+  // after sequential blurs of constant sigma_filter
+  unsigned int s = 0;
+  float radius = sigma_filter;
+  while(radius < sigma_final)
+  {
+    ++s;
+    radius = sqrtf(sqf(radius) + sqf((float)(1 << s) * sigma_filter));
+  }
+  return s + 1;
+}
+
+inline static void blur_2D_Bspline(const float *const restrict in, float *const restrict HF,
+                                   float *const restrict LF,
+                                   const size_t width, const size_t height, const int mult)
+{
+  // Blur and compute the decimated wavelet at once
+#ifdef _OPENMP
+#pragma omp parallel for default(none) dt_omp_firstprivate(width, height, in, LF, HF, mult) \
+    schedule(simd: static)    \
+    collapse(2)
+#endif
+  for(size_t i = 0; i < height; i++)
+  {
+    for(size_t j = 0; j < width; j++)
+    {
+      const size_t index = (i * width + j) * 4; // full scale
+      float DT_ALIGNED_PIXEL acc[4] = { 0.f };
+
+      for(size_t ii = 0; ii < FSIZE; ++ii)
+        for(size_t jj = 0; jj < FSIZE; ++jj)
+        {
+          const size_t row = CLAMP((int)i + mult * (int)(ii - (FSIZE - 1) / 2), (int)0, (int)height - 1);
+          const size_t col = CLAMP((int)j + mult * (int)(jj - (FSIZE - 1) / 2), (int)0, (int)width - 1);
+          const size_t k_index = (row * width + col) * 4;
+
+          const float DT_ALIGNED_ARRAY filter[FSIZE]
+              = { 1.0f / 16.0f, 4.0f / 16.0f, 6.0f / 16.0f, 4.0f / 16.0f, 1.0f / 16.0f };
+          const float filters = filter[ii] * filter[jj];
+
+          for_four_channels(c, aligned(in : 64) aligned(acc : 16)) acc[c] += filters * in[k_index + c];
+        }
+
+      for_four_channels(c, aligned(in, HF, LF : 64) aligned(acc : 16))
```

Comment:
I kind of remember that for_four_channels() was not OK when inside an already SIMD loop, no?

Reply:
The parent loop is not SIMD, no.

Reply:
Oh, I've been mistaken by the fact that you have a `schedule(simd: static)` directive and now I'm wondering if this has some semantics if the loop is not simd? Can you clarify? This is out of my knowledge :)

Reply:
SIMD loops have an explicit `for parallel simd` statement. The `schedule()` statement is only defining how the parallelization should split the loop between threads. `static:simd` means the chunks have the same size as the native SIMD vector, to allow in-thread vectorization when possible.